core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/QC.java

```diff
@@ -23,79 +23,123 @@
- * {@link RunQC}
- *
- * @author Rob Davey
- * @since 0.0.2
- */
-public interface QC extends Securable, Comparable, Deletable {
-  public long getId();
-
-  public void setId(long id);
-
-  /**
-   * Returns the qcId of this QC object.
-   *
-   * @return Long qcId.
-   */
-  @Deprecated
-  public Long getQcId();
-
-  /**
-   * Sets the qcId of this QC object.
-   *
-   * @param qcId qcId.
-   */
-  @Deprecated
-  public void setQcId(Long qcId);
-
-  /**
-   * Returns the qcCreator of this QC object.
-   *
-   * @return String qcCreator.
-   */
-  public String getQcCreator();
-
-  /**
-   * Sets the qcCreator of this QC object.
-   *
-   * @param creator qcCreator.
-   */
-  public void setQcCreator(String creator);
-
-  /**
-   * Returns the qcMethod of this QC object.
-   *
-   * @return String qcMethod.
-   */
-  public QcType getQcType();
-
-  /**
-   * Sets the QcType of this QC object.
-   *
-   * @param type type.
-   */
-  public void setQcType(QcType type);
-
-  /**
-   * Returns the qcDate of this QC object.
-   *
-   * @return Date qcDate.
-   */
-  public Date getQcDate();
-
-  /**
-   * Sets the qcDate of this QC object.
-   *
-   * @param date qcDate.
-   */
-  public void setQcDate(Date date);
+  public static final Long UNSAVED_ID = 0L;
+
+  @ManyToOne(targetEntity = UserImpl.class)
+  @JoinColumn(name = "creator")
+  private User creator;
+
+  @Temporal(TemporalType.DATE)
+  private Date date = new Date();
+
+  @Id
+  @GeneratedValue(strategy = GenerationType.AUTO)
+  private long qcId = QC.UNSAVED_ID;
+
+  private Double results;
+
+  @ManyToOne
+  @JoinColumn(name = "type")
+  private QcType type;
+
+  public User getCreator() {
+    return creator;
+  }
+
+  public Date getDate() {
+    return date;
+  }
+
+  public abstract QualityControlEntity getEntity();
+
+  public long getId() {
+    return qcId;
+  }
+
+  public Double getResults() {
+    return results;
+  }
+
+  public QcType getType() {
+    return type;
+  }
+
+  public void setCreator(User creator) {
+    this.creator = creator;
+  }
+
+  public void setDate(Date date) {
+    this.date = date;
+  }
+
+  public void setId(long qcId) {
+    this.qcId = qcId;
+  }
+
+  public void setResults(Double results) {
+    this.results = results;
+  }
+
+  public void setType(QcType type) {
+    this.type = type;
+  }
+
+  @Override
+  public int compareTo(QC o) {
+    if (type != null && !type.equals(o.getType())) {
+      return type.compareTo(o.getType());
+    }
+    return date.compareTo(o.getDate());
+  }
+
+  @Override
+  public int hashCode() {
+    final int prime = 31;
+    int result = 1;
+    result = prime * result + ((date == null) ? 0 : date.hashCode());
+    result = prime * result + (int) (qcId ^ (qcId >>> 32));
+    result = prime * result + ((results == null) ? 0 : results.hashCode());
+    result = prime * result + ((type == null) ? 0 : type.hashCode());
+    return result;
+  }
+
+  @Override
+  public boolean equals(Object obj) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)